### PR TITLE
test cases/linuxlike/14 static dynamic linkage: Adding matrix by static

### PR DESCRIPTION
**Pull request by dememax**
Not only Solaris doesn't ship static libraries, it's a common practice. See, for example 'test cases/frameworks/1 boost'.

Adding possibility to skip the static linkage with the help of SKIP_STATIC_ZLIB env var (like SKIP_STATIC_BOOST).

Adding the possibility to test dynamic llinkage (it was only static one).

TODO: dynamic test on cygwin

### CODE REVIEW

**Comment by dememax** [comment]
https://github.com/mesonbuild/meson/actions/runs/11594506125/job/32280901101#step:5:587
```
Total failed tests:  1

All failures:
  -> common: 273 both libraries    (default_library=both default_both_libraries=static use_dep=True)
```
Is is my fault?

**Comment by dememax** [comment]
During "cygwin / gccx64ninja (pull_request) Failing", with the help of debug print, I've found:

For dynamic linkage (static=false), nm output in "linuxlike / static dynamic linkage" filtered by "zlibVersion" is:
```
0000000100408150 I __imp_zlibVersion
00000001004010d0 T zlibVersion
```

For static, it doesn't contain the symbol with "__imp_".

**Comment by eli-schwartz** [comment]
Small nit: commit messages should use imperative mood: https://cbea.ms/git-commit/#imperative

**Comment by dememax** [comment]
@eli-schwartz Done! Thanks!

File: test cases/linuxlike/14 static dynamic linkage/meson.build [hunk 1]
```diff
@@ -1,36 +1,30 @@
...
+# The nm utility is then used to check for the existence of this symbol
+# in the resulting binary.
 
-# Solaris does not ship static libraries
-if host_machine.system() == 'sunos'
-  has_static = false
-else
-  has_static = true
-endif
+project('zlib linkage', 'c')
```
[the remark this answers is not in the document]
Comment (dememax) [reply test cases/linuxlike/14 static dynamic linkage/meson.build:6]: Shall be fixed! Thanks.

File: test cases/linuxlike/14 static dynamic linkage/meson.build [hunk 2]
```diff
@@ -1,36 +1,30 @@
-project('static dynamic', 'c')
+# This test uses zlib to verify static and dynamic linkages.
+# A C program is used to call a single zlib function (zlibVersion).
+# The nm utility is then used to check for the existence of this symbol
+# in the resulting binary.
 
-# Solaris does not ship static libraries
-if host_machine.system() == 'sunos'
```
[the remark this answers is not in the document]
Comment (dememax) [reply test cases/linuxlike/14 static dynamic linkage/meson.build:4]: You are perfectly right that this changes the behavior, and it's better to keep the previous logic if we don't want to break something. Nice catch!

I will do what you are asking, but let's go a little farther...

### When my new approach cause test failure

As far as I know, there are **no SunOS tests** in the current CI. And, we both know, **the green status** is showing this.

So there should be some test cases **outside** the project, the dark matter, where people will have this issue.

### When my new approach fixes problem where it could possibly cause test failure

AFAIK, SunOS **could be** configured to have the _static_ zlib version in addition to the default dynamic one. In this case, keeping the old condition is a bug.

Let's ask the author of those lines @alanc 

With my approach the user of this test case will be able to change his environment to adopt the usage on SunOS for his/her setup.

### Symmetry / Special case

What is **so special** about SunOS to include it into this build script?

What about the default configuration on MS Windows (no Cygwin) or MacOS?

What about IBM i (AS/400) and BSD families?

Gentoo setup is configured for Boost with the help of SKIP_STATIC_BOOST in the GHA yaml file. Why could  not SunOS be configured in **the same way**?

### Bottom line

As I said, I will fix this to keep this special case for SunOS, but IMHO it must go.

[the remark this answers is not in the document]
Comment (alanc) [reply test cases/linuxlike/14 static dynamic linkage/meson.build:4]: While there is no SunOS (Solaris or illumos) instance in the CI, we run the test suite on Solaris when building the meson packages for Solaris.  It is true that it is not impossible for a user to build and install a static version of libz on Solaris - but since we provide a dynamic library for libz as part of the OS packages, I don't know of anyone who bothers to build their own static version.  I can't speak to other OS'es or why they didn't hit the problem - perhaps their default configuration doesn't have dynamic libz either instead of just dynamic and not static.

I'm not opposed to moving this from a line of code to a configuration file - what's the "GHA yaml file"?

[the remark this answers is not in the document]
Comment (eli-schwartz) [reply test cases/linuxlike/14 static dynamic linkage/meson.build:4]:
```
$ git --no-pager grep -C3 STATIC_BOOST
.github/workflows/cygwin.yml-        env:
.github/workflows/cygwin.yml-          # Cygwin's static boost installation is broken (some static library
.github/workflows/cygwin.yml-          # variants such as boost_thread are not present)
.github/workflows/cygwin.yml:          SKIP_STATIC_BOOST: 1
.github/workflows/cygwin.yml-        shell: C:\cygwin\bin\bash.exe --noprofile --norc -o igncr -eo pipefail '{0}'
.github/workflows/cygwin.yml-
.github/workflows/cygwin.yml-      - uses: actions/upload-artifact@v4
--
ci/ciimage/fedora/image.json-  "base_image": "fedora:latest",
ci/ciimage/fedora/image.json-  "env": {
ci/ciimage/fedora/image.json-    "CI":                "1",
ci/ciimage/fedora/image.json:    "SKIP_STATIC_BOOST": "1",
ci/ciimage/fedora/image.json-    "MESON_CI_JOBNAME": "linux-fedora-gcc"
ci/ciimage/fedora/image.json-  }
ci/ciimage/fedora/image.json-}
--
ci/ciimage/gentoo/image.json-  "env": {
ci/ciimage/gentoo/image.json-    "CI": "1",
ci/ciimage/gentoo/image.json-    "MESON_CI_JOBNAME": "linux-gentoo-gcc",
ci/ciimage/gentoo/image.json:    "SKIP_STATIC_BOOST": "1"
ci/ciimage/gentoo/image.json-  }
ci/ciimage/gentoo/image.json-}
--
ci/ciimage/opensuse/image.json-  "base_image": "opensuse/tumbleweed:latest",
ci/ciimage/opensuse/image.json-  "env": {
ci/ciimage/opensuse/image.json-    "CI":                  "1",
ci/ciimage/opensuse/image.json:    "SKIP_STATIC_BOOST":   "1",
ci/ciimage/opensuse/image.json-    "SINGLE_DUB_COMPILER": "1",
ci/ciimage/opensuse/image.json-    "MESON_CI_JOBNAME": "linux-opensuse-gcc"
ci/ciimage/opensuse/image.json-  }
--
test cases/frameworks/1 boost/test.json-  "matrix": {
test cases/frameworks/1 boost/test.json-    "options": {
test cases/frameworks/1 boost/test.json-      "static": [
test cases/frameworks/1 boost/test.json:        { "val": "true", "skip_on_env": [ "SKIP_STATIC_BOOST" ] },
test cases/frameworks/1 boost/test.json-        { "val": "false" }
test cases/frameworks/1 boost/test.json-      ],
test cases/frameworks/1 boost/test.json-      "b_vscrt": [
```
For all our docker-based CI testing we generate the docker images themselves with `/ci/env_vars.sh` as a kind of "/etc/profile for CI".

For Cygwin, which isn't docker-based, we use the GitHub Actions (GHA) yaml file `.github/workflows/cygwin.yml`. We don't maintain any such thing for Solaris' packaging.

[the remark this answers is not in the document]
Comment (dememax) [reply test cases/linuxlike/14 static dynamic linkage/meson.build:4]: Thank you, @alanc !
So, should we throw away the SunOS special condition, or merge the PR in the current state?

[the remark this answers is not in the document]
Comment (dememax) [reply test cases/linuxlike/14 static dynamic linkage/meson.build:4]: Excuse me, @eli-schwartz, what do you think?

**Comment by dememax** [comment]
Only one fatal error not caused by my changes:
```
Ubuntu Bionic
failed Dec 3, 2024 in 46s

Run actions/checkout@v3
/usr/bin/docker exec  7f49c64d95a49a1cc1327eb78b5f56bc3e01d8e7f8e38f1f2dcc91e71e0b1614 sh -c "cat /etc/*release | grep ^ID"
/__e/node20/bin/node: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by /__e/node20/bin/node)
```
https://github.com/mesonbuild/meson/actions/runs/12135619615/job/33835043547?pr=13842
